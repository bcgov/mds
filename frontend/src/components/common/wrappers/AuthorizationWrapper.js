import React from "react";
import { PropTypes } from "prop-types";
import { connect } from "react-redux";
import { getUserAccessData } from "@/selectors/authenticationSelectors";
import { USER_ROLES } from "@/constants/environment";

/**
 * @constant AuthorizationWrapper conditionally renders react children depending
 * if the role passed in matches the user permissions.
 * 
 * NOTE: If the childComponent is an ant design child component**
 * IE. Menu.Item,
 * <Menu>
 *    <Menu.Item />
 * </Menu>
 *  The implicit flow of passing props down to children could be interrupted.
 * Ie, Menu passes onItemHover down to Menu.Item, with AuthorizationWrapper wrapping Menu.Item, it no longer has access
 * to onItemHover, as a solution, a <div className="custom-menu-item" /> will be used in place of Menu.Item and will inherit all CSS from 
 * Menu.Item.
 * NavBar.js use case example: 
 * <Menu>
    <AuthorizationWrapper permission="role_admin">
      <div className="custom-menu-item">
        <Link to={router.ADMIN_DASHBOARD.route}>
          Admin
        </Link>
      </div>
    </AuthorizationWrapper>
  </Menu>
 * 
 * NOTE: isMajorMine comes from `mine.major_mine_ind`, currently in MDS only Major mines can be updated, 
 * therefore all eidt buttons will be hidden from regional Mines.
 */

const propTypes = {
  permission: PropTypes.string.isRequired,
  isMajorMine: PropTypes.bool,
  children: PropTypes.element.isRequired,
};

const defaultProps = {
  isMajorMine: true,
};
export const AuthorizationWrapper = (props) =>
  props.userRoles.includes(USER_ROLES[props.permission]) &&
  props.isMajorMine && <div>{...props.children}</div>;

AuthorizationWrapper.propTypes = propTypes;
AuthorizationWrapper.defaultProps = defaultProps;

const mapStateToProps = (state) => ({
  userRoles: getUserAccessData(state),
});

export default connect(mapStateToProps)(AuthorizationWrapper);
