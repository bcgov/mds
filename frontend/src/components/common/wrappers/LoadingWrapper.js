import React from "react";
import { PropTypes } from "prop-types";
import { Spin, Icon } from "antd";
import { TABLE_SKELETON } from "@/constants/assets";
import * as Style from "@/constants/styles";

/**
 * @constant LoadingWrapper renders react children || or loading view
 */

const propTypes = {
  condition: PropTypes.bool.isRequired,
  children: PropTypes.element.isRequired,
  type: PropTypes.oneOf(["spinner", "table-skeleton"]),
};

const defaultProps = {
  type: "spinner",
};

export const LoadingWrapper = (props) => {
  const antIcon = (
    <Icon type="loading" style={{ fontSize: 80, color: Style.COLOR.mediumGrey }} spin />
  );

  const renderLoading = (
    <div className="page__content">
      {props.type === "table-skeleton" && (
        <img src={TABLE_SKELETON} alt="loading" id="table-skeleton" />
      )}
      {props.type === "spinner" && (
        <div className="loading-screen--small">
          <Spin id="spinner" indicator={antIcon} />
        </div>
      )}
    </div>
  );

  return (
    <div>
      {props.condition ? (
        <div className="fade-in">{props.children}</div>
      ) : (
        <div>{renderLoading}</div>
      )}
    </div>
  );
};

LoadingWrapper.propTypes = propTypes;
LoadingWrapper.defaultProps = defaultProps;

export default LoadingWrapper;
