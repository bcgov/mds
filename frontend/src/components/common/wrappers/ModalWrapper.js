import React, { Component } from "react";
import { bindActionCreators } from "redux";
import LoadingBar from "react-redux-loading-bar";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { Modal, Icon, Button, Popconfirm } from "antd";
import * as Styles from "@/constants/styles";
import { closeModal } from "@/actions/modalActions";
import {
  getIsModalOpen,
  getProps,
  getContent,
  getClearOnSubmit,
  getWidthSize,
} from "@/selectors/modalSelectors";

const propTypes = {
  closeModal: PropTypes.func.isRequired,
  isModalOpen: PropTypes.bool.isRequired,
  content: PropTypes.func,
  props: PropTypes.objectOf(PropTypes.string),
  clearOnSubmit: PropTypes.bool.isRequired,
  widthSize: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
};

const defaultProps = {
  props: {
    title: "",
  },
  content: () => {},
};

export class ModalWrapper extends Component {
  constructor(props) {
    super(props);
    // listens for browser back || forward button click and invokes function to close the modal,
    window.onpopstate = this.onBrowserButtonEvent;
  }

  onBrowserButtonEvent = () => {
    this.props.closeModal();
  };

  renderTitle = () => (
    <div className="inline-flex between">
      <h1>{this.props.props.title}</h1>
      <Icon type="close" />
    </div>
  );

  closeModal = (event) => {
    event.preventDefault();
    this.props.closeModal();
  };

  render() {
    const ChildComponent = this.props.content;
    return (
      <Modal
        width={this.props.widthSize}
        title={this.props.props.title}
        visible={this.props.isModalOpen}
        closable={false}
        footer={null}
      >
        <Popconfirm
          placement="bottomRight"
          title="Are you sure you want to cancel?"
          okText="Yes"
          cancelText="No"
          onConfirm={(event) => this.closeModal(event)}
        >
          <Button ghost className="modal__close">
            <Icon type="close" />
          </Button>
        </Popconfirm>
        <LoadingBar
          scope="modal"
          style={{
            position: "absolute",
            top: "50px",
            left: 0,
            backgroundColor: Styles.COLOR.violet,
            width: "100%",
            height: "8px",
            zIndex: 100,
          }}
        />
        {ChildComponent && (
          <ChildComponent
            closeModal={this.props.closeModal}
            clearOnSubmit={this.props.clearOnSubmit}
            {...this.props.props}
          />
        )}
      </Modal>
    );
  }
}

const mapStateToProps = (state) => ({
  widthSize: getWidthSize(state),
  isModalOpen: getIsModalOpen(state),
  props: getProps(state),
  content: getContent(state),
  clearOnSubmit: getClearOnSubmit(state),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      closeModal,
    },
    dispatch
  );

ModalWrapper.propTypes = propTypes;
ModalWrapper.defaultProps = defaultProps;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ModalWrapper);
