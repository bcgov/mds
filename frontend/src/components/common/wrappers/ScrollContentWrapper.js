import React from "react";
import { withRouter } from "react-router-dom";
import { includes } from "lodash";
import { PropTypes } from "prop-types";

/**
 * @constant ScrollContentWrapper renders react children with an active indicator if the id is in the url.
 */

const propTypes = {
  id: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired,
  location: PropTypes.shape({ hash: PropTypes.string }).isRequired,
};

export const ScrollContentWrapper = (props) => {
  const isActive = () =>
    includes(props.id, props.location.hash) ? "circle--purple" : "circle--gray";

  return (
    <div className="scroll">
      <div className={isActive()} />
      <div>{props.children}</div>
    </div>
  );
};

ScrollContentWrapper.propTypes = propTypes;

export default withRouter(ScrollContentWrapper);
