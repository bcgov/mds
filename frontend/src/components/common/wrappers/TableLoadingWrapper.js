import React from "react";
import { PropTypes } from "prop-types";
import { Table } from "antd";

/**
 * @constant TableLoadingWrapper renders react children or a skeleton loading view using the column headers
 * of the tables and 9 loading rows
 *
 * condition = expecting a truthy value to render children, ie "isLoaded"
 * tableHeaders = array of table headers, the skeleton loading will have the same number of columns and same titles
 * isPaginated = true if the table is paginated - this allows the skeleton to be smooth when changing pages
 *
 *
 */

const propTypes = {
  condition: PropTypes.bool.isRequired,
  children: PropTypes.element.isRequired,
  tableHeaders: PropTypes.arrayOf(PropTypes.objectOf(PropTypes.strings)).isRequired,
  isPaginated: PropTypes.bool,
};

const defaultProps = {
  isPaginated: false,
};

export const TableLoadingWrapper = (props) => {
  const renderColumns = props.tableHeaders.map((title) => ({
    title,
    dataIndex: title,
    width: 150,
    render: () => <div className="skeleton-table__loader" />,
  }));
  return (
    <div>
      {props.condition ? (
        <div>{props.children}</div>
      ) : (
        <div className="skeleton-table">
          <Table
            align="left"
            pagination={false}
            columns={renderColumns}
            dataSource={new Array(props.isPaginated ? 25 : 9).fill({})}
            rowClassName="skeleton-table__row"
          />
        </div>
      )}
    </div>
  );
};

TableLoadingWrapper.propTypes = propTypes;
TableLoadingWrapper.defaultProps = defaultProps;

export default TableLoadingWrapper;
