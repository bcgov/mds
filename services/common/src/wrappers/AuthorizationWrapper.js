import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { startCase, camelCase } from "lodash";
import { USER_ROLES, detectDevelopmentEnvironment, detectProdEnvironment } from "@mds/common";
import { Tooltip } from "antd";
import { useSelector } from "react-redux";
import { userHasRole } from "@mds/common/redux/reducers/authenticationReducer";

/**
 * @constant AuthorizationWrapper conditionally renders react children depending
 * if the role passed in matches the user permissions.
 *
 * NOTE: If the childComponent is an ant design child component**
 * IE. Menu.Item,
 * <Menu>
 *    <Menu.Item />
 * </Menu>
 *  The implicit flow of passing props down to children could be interrupted.
 * Ie, Menu passes onItemHover down to Menu.Item, with AuthorizationWrapper wrapping Menu.Item, it no longer has access
 * to onItemHover, as a solution, a <div className="custom-menu-item" /> will be used in place of Menu.Item and will inherit all CSS from
 * Menu.Item.
 * NavBar.js use case example:
 * <Menu>
    <AuthorizationWrapper permission="role_admin">
      <div className="custom-menu-item">
        <Link to={router.ADMIN_DASHBOARD.route}>
          Admin
        </Link>
      </div>
    </AuthorizationWrapper>
  </Menu>
 *
 * NOTE: isMajorMine comes from `mine.major_mine_ind`, currently in MDS only Major mines can be updated,
 * therefore all edit buttons will be hidden from regional Mines -- Admin can view/edit everything
 *
 *
 * inDevelopment - If the feature is still being built and not ready to be shared with a larger audience, `inDevelopment` only displays the content in local and dev environment
 * inTesting - if the feature is ready to be shared with a larger audience, but not ready to be displayed in PROD, `inTesting` will display content in every environment except Prod.
 */

const propTypes = {
  permission: PropTypes.string,
  isMajorMine: PropTypes.bool,
  inDevelopment: PropTypes.bool,
  inTesting: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element.isRequired),
    PropTypes.element.isRequired,
  ]),
  userRoles: PropTypes.arrayOf(PropTypes.string).isRequired,
  showToolTip: PropTypes.bool,
};

const defaultProps = {
  isMajorMine: undefined,
  inDevelopment: undefined,
  inTesting: undefined,
  permission: undefined,
  showToolTip: true,
  userRoles: [],
};

export const AuthorizationWrapper = (props) => {
  const inDevCheck =
    props.inDevelopment === undefined || (props.inDevelopment && detectDevelopmentEnvironment());
  const inTestCheck =
    props.inTesting === undefined || (props.inTesting && !detectProdEnvironment());
  const permissionCheck = useSelector((state) => userHasRole(state, props.permission));
  const isMajorMine = props.isMajorMine === undefined || props.isMajorMine;
  const isAdmin = useSelector((state) => userHasRole(state, USER_ROLES.role_admin));

  const title = () => {
    const permission = props.permission ? `${USER_ROLES[props.permission]}` : "";
    const inTest = props.inTesting ? "Not Visible in Production" : "";
    const majorMine = props.isMajorMine !== undefined ? "Only Visible to Major Mines" : "";
    return (
      <ul style={{ listStyle: "none", marginBottom: "0" }}>
        {permission && <li>{startCase(camelCase(permission))}</li>}
        {inTest && <li>{inTest}</li>}
        {majorMine && <li>{majorMine}</li>}
      </ul>
    );
  };

  // all actions are visible to admin, except when they should only exist in their respective environments. (ie admin cannot see features in prod if they are feature flagged to test)
  const adminOverride = isAdmin && !inDevCheck && !inTestCheck;
  return (
    (adminOverride || (inDevCheck && inTestCheck && permissionCheck && isMajorMine)) && (
      <Tooltip
        title={isAdmin && props.showToolTip ? title() : ""}
        placement="left"
        mouseEnterDelay={1.7}
        mouseLeaveDelay={0}
        arrowPointAtCenter
        overlayClassName="tooltip__admin"
        style={{ zIndex: 100000 }}
        trigger={["hover"]}
        destroyTooltipOnHide
      >
        {React.createElement("span", null, props.children)}
      </Tooltip>
    )
  );
};
AuthorizationWrapper.propTypes = propTypes;
AuthorizationWrapper.defaultProps = defaultProps;

export default connect()(AuthorizationWrapper);
