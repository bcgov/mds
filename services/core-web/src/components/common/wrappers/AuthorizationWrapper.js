/* eslint-disable */
import React from "react";
import ReactDOMServer from "react-dom/server";
import { PropTypes } from "prop-types";
import { connect } from "react-redux";
import { startCase, camelCase } from "lodash";
import { getUserAccessData } from "@common/selectors/authenticationSelectors";
import { USER_ROLES } from "@common/constants/environment";
import {
  detectDevelopmentEnvironment,
  detectProdEnvironment,
} from "@common/utils/environmentUtils";
import { Tooltip } from "antd";
import * as Permission from "@/constants/permissions";

/**
 * @constant AuthorizationWrapper conditionally renders react children depending
 * if the role passed in matches the user permissions.
 *
 * NOTE: If the childComponent is an ant design child component**
 * IE. Menu.Item,
 * <Menu>
 *    <Menu.Item />
 * </Menu>
 *  The implicit flow of passing props down to children could be interrupted.
 * Ie, Menu passes onItemHover down to Menu.Item, with AuthorizationWrapper wrapping Menu.Item, it no longer has access
 * to onItemHover, as a solution, a <div className="custom-menu-item" /> will be used in place of Menu.Item and will inherit all CSS from
 * Menu.Item.
 * NavBar.js use case example:
 * <Menu>
    <AuthorizationWrapper permission="role_admin">
      <div className="custom-menu-item">
        <Link to={router.ADMIN_DASHBOARD.route}>
          Admin
        </Link>
      </div>
    </AuthorizationWrapper>
  </Menu>
 *
 * NOTE: isMajorMine comes from `mine.major_mine_ind`, currently in MDS only Major mines can be updated,
 * therefore all edit buttons will be hidden from regional Mines -- Admin can view/edit everything
 *
 *
 * inDevelopment - If the feature is still being built and not ready to be shared with a larger audience, `inDevelopment` only displays the content in local and dev environment
 * inTesting - if the feature is ready to be shared with a larger audience, but not ready to be displayed in PROD, `inTesting` will display content in every environment except Prod.
 */

const propTypes = {
  permission: PropTypes.string,
  isMajorMine: PropTypes.bool,
  inDevelopment: PropTypes.bool,
  inTesting: PropTypes.bool,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.element.isRequired),
    PropTypes.element.isRequired,
  ]),
  userRoles: PropTypes.arrayOf(PropTypes.string).isRequired,
};

const defaultProps = {
  isMajorMine: undefined,
  inDevelopment: undefined,
  inTesting: undefined,
  permission: undefined,
};

export const AuthorizationWrapper = (props) => {
  const inDevCheck =
    props.inDevelopment === undefined || (props.inDevelopment && detectDevelopmentEnvironment());
  const inTestCheck =
    props.inTesting === undefined || (props.inTesting && !detectProdEnvironment());
  const permissionCheck =
    props.permission === undefined || props.userRoles.includes(USER_ROLES[props.permission]);
  const isMajorMine = props.isMajorMine === undefined || props.isMajorMine;
  const isAdmin = props.userRoles.includes(USER_ROLES[Permission.ADMIN]);

  const title = () => {
    const permission = props.permission ? `${USER_ROLES[props.permission]}` : "";
    const inTest = props.inTesting ? "Not Visible in Production" : "";
    const majorMine = props.isMajorMine !== undefined ? "Only Visible to Major Mines" : "";
    return (
      <ul style={{ listStyle: "none", marginBottom: "0" }}>
        {permission && <li>{startCase(camelCase(permission))}</li>}
        {inTest && <li>{inTest}</li>}
        {majorMine && <li>{majorMine}</li>}
      </ul>
    );
  };

  return (
    (isAdmin || (inDevCheck && inTestCheck && permissionCheck && isMajorMine)) && (
      <Tooltip
        title={isAdmin ? title() : ""}
        placement="left"
        mouseEnterDelay={1}
        arrowPointAtCenter
        overlayClassName="tooltip__admin"
        style={{ zIndex: 100000 }}
      >
        {React.createElement("span", null, props.children)}
      </Tooltip>
    )
  );
};
AuthorizationWrapper.propTypes = propTypes;
AuthorizationWrapper.defaultProps = defaultProps;

const mapStateToProps = (state) => ({
  userRoles: getUserAccessData(state),
});

export default connect(mapStateToProps)(AuthorizationWrapper);
