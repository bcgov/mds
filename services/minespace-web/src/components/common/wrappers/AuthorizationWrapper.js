import React from "react";
import { PropTypes } from "prop-types";
import { connect } from "react-redux";
import { detectDevelopmentEnvironment, detectProdEnvironment } from "@/utils/environmentUtils";
import { isProponent } from "@/selectors/authenticationSelectors";

/**
 * @constant AuthorizationWrapper conditionally renders react children depending
 * if the role passed in matches the user permissions.
 *
 * NOTE: If the childComponent is an ant design child component**
 * IE. Menu.Item,
 * <Menu>
 *    <Menu.Item />
 * </Menu>
 * The implicit flow of passing props down to children could be interrupted.
 * Ie, Menu passes onItemHover down to Menu.Item, with AuthorizationWrapper wrapping Menu.Item, it no longer has access
 * to onItemHover, as a solution, a <div className="custom-menu-item" /> will be used in place of Menu.Item and will inherit all CSS from
 * Menu.Item.
 * Use case example:
 * <Menu>
    <AuthorizationWrapper inDevelopment={true}>
      <div className="custom-menu-item">
        <Button>
          Dev-Only Button
        </Button>
      </div>
    </AuthorizationWrapper>
  </Menu>
 *
 * inDevelopment - if the feature is still being built and not ready to be shared with a larger audience, `inDevelopment` only displays the content in local and dev environment
 * inTesting - if the feature is ready to be shared with a larger audience, but not ready to be displayed in PROD, `inTesting` will display content in every environment except Prod.
 * isProponent - This prop comes directly from the store, only MineSpace proponents have access to create anything, CORE users are view only
 */

const propTypes = {
  inDevelopment: PropTypes.bool,
  inTesting: PropTypes.bool,
  children: PropTypes.element.isRequired,
};

const defaultProps = {
  inDevelopment: undefined,
  inTesting: undefined,
};

export const AuthorizationWrapper = (props) => {
  const checkDev = props.inDevelopment && detectDevelopmentEnvironment();
  const checkTest = props.inTesting && !detectProdEnvironment();
  // do not show any actions if the user is not a proponents, unless in the development
  if (!props.isProponent && !detectDevelopmentEnvironment()) {
    return <span />;
  } else {
    if (props.inDevelopment === undefined && props.inTesting === undefined) {
      return <span>{props.children}</span>;
    } else if (checkDev || checkTest) {
      return <span>{props.children}</span>;
    }
  }
};

const mapStateToProps = (state) => ({
  isProponent: isProponent(state),
});

AuthorizationWrapper.propTypes = propTypes;
AuthorizationWrapper.defaultProps = defaultProps;

export default connect(mapStateToProps)(AuthorizationWrapper);
