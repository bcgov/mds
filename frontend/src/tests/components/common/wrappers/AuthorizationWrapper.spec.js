import React from "react";
import { shallow } from "enzyme";
import { AuthorizationWrapper } from "@/components/common/wrappers/AuthorizationWrapper";
import * as Mock from "@/tests/mocks/dataMocks";

const props = {};

const setupProps = () => {
  props.permission = "role_admin";
  props.isMajorMine = true;
  props.userRoles = Mock.USER_ACCESS_DATA;
  props.children = <div>helo</div>;
};

beforeEach(() => {
  setupProps();
});

describe("AuthorizationWrapper", () => {
  it("renders properly", () => {
    const wrapper = shallow(<AuthorizationWrapper {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
