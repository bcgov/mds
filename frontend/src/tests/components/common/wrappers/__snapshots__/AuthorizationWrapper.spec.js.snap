// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizationWrapper renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthorizationWrapper
    inDevelopment={false}
    inTesting={false}
    isMajorMine={true}
    permission="role_admin"
    userRoles={
      Array [
        "mds-mine-view",
        "idir",
        "mds-mine-create",
        "offline_access",
        "admin",
        "uma_authorization",
        "mds-mine-admin",
      ]
    }
  >
    <div>
      helo
    </div>
  </AuthorizationWrapper>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        false,
        false,
        <span>
          <div>
            helo
          </div>
        </span>,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div>
            helo
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "helo",
          },
          "ref": null,
          "rendered": "helo",
          "type": "div",
        },
        "type": "span",
      },
    ],
    "type": "span",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          false,
          false,
          <span>
            <div>
              helo
            </div>
          </span>,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div>
              helo
            </div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "helo",
            },
            "ref": null,
            "rendered": "helo",
            "type": "div",
          },
          "type": "span",
        },
      ],
      "type": "span",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
